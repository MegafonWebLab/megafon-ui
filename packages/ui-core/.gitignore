.idea
*.iml
node_modules

src/**/*.css
src/index.ts
src/components/**/*.d.ts
dist
coverage
.docz
docs
.vscode
.DS_Store
