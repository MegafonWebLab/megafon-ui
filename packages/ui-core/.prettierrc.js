module.exports = {
    ...require('@megafon/frontend-presets/prettier'),
};
