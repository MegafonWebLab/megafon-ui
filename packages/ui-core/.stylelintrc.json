{
  "rules": {
    "color-hex-case": "upper",
    "color-no-invalid-hex": true,
    "font-family-name-quotes": "always-where-recommended",
    "function-calc-no-unspaced-operator": true,
    "function-comma-space-after": "always",
    "function-comma-space-before": "never",
    "number-leading-zero": "always",
    "number-no-trailing-zeros": true,
    "string-no-newline": true,
    "string-quotes": "single",
    "length-zero-no-unit": true,
    "unit-case": "lower",
    "unit-no-unknown": true,
    "value-keyword-case": "lower",
    "value-no-vendor-prefix": true,
    "property-case": "lower",
    "property-no-unknown": true,
    "keyframe-declaration-no-important": true,
    "declaration-bang-space-after": "never",
    "declaration-block-no-duplicate-properties": true,
    "declaration-block-trailing-semicolon": "always",
    "declaration-block-semicolon-space-before": "never",
    "block-no-empty": true,
    "block-opening-brace-newline-after": "always",
    "block-closing-brace-newline-before": "always",
    "selector-attribute-brackets-space-inside": "never",
    "max-nesting-depth": 5,
    "indentation": 4,
    "declaration-bang-space-before": "always"
  }
}
