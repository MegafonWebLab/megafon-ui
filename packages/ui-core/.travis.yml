sudo: false
language: node_js

node_js:
  - 10

cache:
  directories:
  - node_modules

before_script:
  - npm install -g codecov
  - npm ci

script:
  - npm run lint
  - npm run coverage

before_deploy:
  - npm run docz:build

deploy:
  - provider: npm
    email: "$NPM_EMAIL"
    api_key: "$NPM_TOKEN"
    skip_cleanup: true
    on:
      tags: true
      node_js: "10"

  - provider: pages
    github-token: "$GITHUB_TOKEN"
    skip-cleanup: true
    local-dir: "docs"
    keep-history: true
    on:
      branch: master
      node_js: "10"

after_script:
  - codecov
