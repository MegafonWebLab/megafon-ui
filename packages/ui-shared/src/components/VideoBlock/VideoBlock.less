@import '~@megafon/ui-core/styles/base.less';
@b: e('.mfui-beta-video-block');

@{b} {
    display: flex;
    align-items: center;

    &__content {
        display: flex;
        flex-direction: column;
        height: 100%;
        justify-content: center;

        @media @mobileSM {
            margin-top: 24px;
        }

        @media @mobileB {
            margin-top: 32px;
        }

        @media @mobileBU {
            padding-right: 20px;
            margin-right: 20px;
        }
    }

    &__header {
        margin-bottom: 24px;

        @media @mobile {
            margin-bottom: 16px;
        }
    }

    &__text:not(:first-child) {
        margin-top: 24px;
    }

    &__button {
        margin-top: 32px;
        width: 171px;

        @media @mobileSM {
            align-self: center;
        }
    }

    &__video {
        width: 100%;
        height: 100%;
        object-fit: contain;
    }

    &__video-wrapper {
        height: 674px;
        overflow: hidden;

        @media @desktopB {
            height: 600px;
        }

        @media @desktopM {
            height: 524px;
        }

        @media @desktopS {
            height: 505px;
        }
    }

    &__video-wrapper_with-content {
        height: 468px;

        @media @desktopB {
            height: 415px;
        }

        @media @desktopM {
            height: 363px;
        }

        @media @desktopS {
            height: 328px;
        }
    }

    &__video-wrapper,
    &__video-wrapper_with-content {
        @media @mobileB {
            height: 379px;
        }

        @media @mobileSM {
            height: 162px;
        }
    }

    &__grid {
        width: 100%;
    }
}
