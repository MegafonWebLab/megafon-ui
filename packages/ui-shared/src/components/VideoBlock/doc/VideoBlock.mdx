---
name: VideoBlock
menu: Конструктор
route: components/video-block
---

import Example from './VideoBlock.example.mdx';
import Props from './VideoBlock.props.mdx';

# VideoBlock

<CopyToClipboardBox text="import { VideoBlock } from '@megafon/ui-shared';" />

<DoczTabs>
    <Example />
    <Props />
</DoczTabs>
