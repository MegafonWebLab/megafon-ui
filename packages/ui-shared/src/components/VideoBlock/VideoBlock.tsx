import React, { Ref } from 'react';
import PropTypes from 'prop-types';
import './VideoBlock.less';
import { Header, Button, Paragraph, Grid, GridColumn, cnCreate, dataAttrs as filterDataAttrs } from '@megafon/ui-core';

export interface IContent {
    /** Заголовок */
    title: string;
    /** Текст-описание */
    description: string[];
    /** Текст кнопки */
    buttonTitle: string;
    /** Ссылка на кнопке */
    href?: string;
    /** Обработчик клика по кнопке */
    onButtonClick?: (e: React.SyntheticEvent<EventTarget>) => void;
}

export const VideoTypes = {
    YOUTUBE: 'youtube',
    VIDEO: 'video',
} as const;

type VideoType = typeof VideoTypes[keyof typeof VideoTypes];

export interface Props {
    /** Дата атрибуты для корневого элемента */
    dataAttrs?: { [key: string]: string };
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Дополнительные классы для корневого и внутренних элементов */
    classes?: {
        root?: string;
        button?: string;
    };
    /** Ссылка на корневой элемент */
    rootRef?: Ref<HTMLDivElement>;
    /** Данные для блока с контентом */
    content?: IContent;
    /** Тип видео */
    videoType?: VideoType;
    /** Источник видео. Если видео с youtube, то необходимо указать id */
    videoSrc: string;
    /** Наличие звука в видео */
    isMuted?: boolean;
    /** Автоматическое проигрывание видео */
    isAutoplay?: boolean;
}

const cn = cnCreate('mfui-beta-video-block');
const VideoBlock: React.FC<Props> = ({
    dataAttrs,
    className,
    classes = {},
    rootRef,
    content,
    videoType = 'video',
    videoSrc,
    isMuted = true,
    isAutoplay = false,
}) => {
    const renderVideo = React.useCallback(() => {
        switch (videoType) {
            case(VideoTypes.YOUTUBE): {
                const src = `https://www.youtube.com/embed/${videoSrc}?&autoplay=${isAutoplay ? 1 : 0}&mute=${isMuted ? 1 : 0}&loop=1&rel=0&controls=0&showinfo=0e&iv_load_policy=3&playlist=${videoSrc}`;

                return (
                    <iframe src={src} width="100%" height="100%" frameBorder="0" allow="autoplay"/>
                );
            }

            case (VideoTypes.VIDEO): {
                return (
                    <video className={cn('video')} autoPlay={isAutoplay} muted={isMuted} controls={!isAutoplay} loop>
                        <source src={videoSrc} type="video/mp4" />
                    </video>
                );
            }

            default: {
                return null;
            }
        }

    }, [videoType, videoSrc]);

    const renderContent = React.useCallback((data: IContent) => {
        const { title, description, href, buttonTitle, onButtonClick } = data;

        return (
            <div className={cn('content')}>
                <Header as="h3" className={cn('header')}>
                    {title}
                </Header>
                <div>
                    {description && description.map((paragraph, i) => (
                        <Paragraph key={i + paragraph} className={cn('text')} hasMargin={false}>
                            {paragraph}
                        </Paragraph>
                    ))}
                </div>
                <Button className={cn('button', [classes.button])} href={href} onClick={onButtonClick}>
                    {buttonTitle}
                </Button>
            </div>
        );
    }, [content]);

    const renderGridColumns = React.useCallback(() => {
        const columns: JSX.Element[] = [];
        const columnWidth = content ? '7' : '10';

        if (content) {
            columns.push(
                <GridColumn all="5" tablet="12" mobile="12" orderTablet="2" orderMobile="2" key={'column-content'}>
                    {renderContent && renderContent(content)}
                </GridColumn>
            );
        }

        columns.push(
            <GridColumn all={columnWidth} tablet="12" mobile="12" key={'column-video'}>
                <div className={cn('video-wrapper', {'with-content': !!content })}>
                    {renderVideo()}
                </div>
            </GridColumn>
        );

        return columns;
    }, [renderContent, renderVideo, content]);

    return (
        <div {...filterDataAttrs(dataAttrs)} className={cn([className, classes.root])} ref={rootRef}>
            <Grid hAlign="center" className={cn('grid')}>
                {renderGridColumns()}
            </Grid>
        </div>
    );
};

VideoBlock.propTypes = {
    dataAttrs: PropTypes.objectOf(PropTypes.string.isRequired),
    className: PropTypes.string,
    classes: PropTypes.shape({
        root: PropTypes.string,
        button: PropTypes.string,
    }),
    rootRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.oneOfType([PropTypes.shape({ current: PropTypes.elementType }), PropTypes.any ]),
    ]),
    content: PropTypes.shape({
        title: PropTypes.string.isRequired,
        description: PropTypes.array.isRequired,
        href: PropTypes.string,
        buttonTitle: PropTypes.string.isRequired,
        onButtonClick: PropTypes.func,
    }),
    videoType: PropTypes.oneOf(Object.values(VideoTypes)),
    videoSrc: PropTypes.string.isRequired,
    isMuted: PropTypes.bool,
    isAutoplay: PropTypes.bool,
};

export default VideoBlock;
