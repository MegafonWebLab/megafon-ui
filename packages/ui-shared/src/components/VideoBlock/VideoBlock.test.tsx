import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { cnCreate } from '@megafon/ui-core';
import VideoBlock, { VideoTypes } from './VideoBlock';

const content = {
    title: 'Test title',
    description: [
        'Test description',
        'Test description2',
    ],
    buttonTitle: 'Button title',
    href: '#',
};

const cn = cnCreate('mfui-beta-video-block');
describe('<VideoBlock />', () => {
    it('it renders VideoBlock with default props', () => {
        const component = shallow(<VideoBlock videoSrc="video.mp4" />);
        expect(component).toMatchSnapshot();
    });

    it('it renders VideoBlock with content', () => {
        const component = shallow(
            <VideoBlock videoSrc="video.mp4" content={content} />
        );
        expect(component).toMatchSnapshot();
    });

    it('it renders VideoBlock with classes', () => {
        const component = shallow(
            <VideoBlock
                videoSrc="video.mp4"
                content={content}
                className="className"
                classes={{ root: 'rootClass', button: 'buttonClass' }} />
        );
        expect(component).toMatchSnapshot();
    });

    it('it renders VideoBlock with dataAttrs', () => {
        const component = shallow(
            <VideoBlock
                videoSrc="video.mp4"
                dataAttrs={{ 'data-test': 'value' }}
            />
        );
        expect(component.first().prop('data-test')).toEqual('value');
    });

    it('it renders VideoBlock with sound turned on', () => {
        const component = shallow(
            <VideoBlock videoSrc="video" isMuted={false} />
        );
        expect(component).toMatchSnapshot();
    });

    it('it renders VideoBlock with autoplay', () => {
        const component = shallow(
            <VideoBlock videoSrc="video" isAutoplay />
        );
        expect(component).toMatchSnapshot();
    });

    it('it renders VideoBlock with youtube media type', () => {
        const component = shallow(
            <VideoBlock videoSrc="youtube" videoType={VideoTypes.YOUTUBE} />
        );
        expect(component).toMatchSnapshot();
    });

    it('it renders VideoBlock with youtube media type and sound turned on', () => {
        const component = shallow(
            <VideoBlock videoSrc="youtube" videoType={VideoTypes.YOUTUBE} isMuted={false}/>
        );
        expect(component).toMatchSnapshot();
    });

    it('it renders VideoBlock with youtube media type and autoplay', () => {
        const component = shallow(
            <VideoBlock videoSrc="youtube" videoType={VideoTypes.YOUTUBE} isAutoplay />
        );
        expect(component).toMatchSnapshot();
    });

    it('should return reference to root element', () => {
        const ref: React.RefObject<HTMLDivElement> = React.createRef();

        mount(<VideoBlock videoSrc="video" isMuted={false} rootRef={ref} />);

        expect(ref.current).not.toBeNull();
    });

    it('should call onClick props', () => {
        const onButtonClick = jest.fn();
        const contentWithMockFunc = {
            ...content,
            onButtonClick,
        };
        const component = shallow(
            <VideoBlock videoSrc="video.mp4" content={contentWithMockFunc} />
        );
        const btn = component.find(`.${cn('button')}`);

        btn.simulate('click');
        expect(onButtonClick).toBeCalled();
    });
});
