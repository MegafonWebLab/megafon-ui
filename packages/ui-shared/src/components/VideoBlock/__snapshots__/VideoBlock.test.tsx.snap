// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoBlock /> it renders VideoBlock with autoplay 1`] = `
<div
  className="mfui-beta-video-block"
>
  <Grid
    className="mfui-beta-video-block__grid"
    hAlign="center"
  >
    <GridColumn
      all="10"
      key="column-video"
      mobile="12"
      tablet="12"
    >
      <div
        className="mfui-beta-video-block__video-wrapper"
      >
        <video
          autoPlay={true}
          className="mfui-beta-video-block__video"
          controls={false}
          loop={true}
          muted={true}
        >
          <source
            src="video"
            type="video/mp4"
          />
        </video>
      </div>
    </GridColumn>
  </Grid>
</div>
`;

exports[`<VideoBlock /> it renders VideoBlock with classes 1`] = `
<div
  className="mfui-beta-video-block className rootClass"
>
  <Grid
    className="mfui-beta-video-block__grid"
    hAlign="center"
  >
    <GridColumn
      all="5"
      key="column-content"
      mobile="12"
      orderMobile="2"
      orderTablet="2"
      tablet="12"
    >
      <div
        className="mfui-beta-video-block__content"
      >
        <Header
          as="h3"
          className="mfui-beta-video-block__header"
          color="black"
          hAlign="inherit"
        >
          Test title
        </Header>
        <div>
          <Paragraph
            className="mfui-beta-video-block__text"
            hasMargin={false}
            key="0Test description"
          >
            Test description
          </Paragraph>
          <Paragraph
            className="mfui-beta-video-block__text"
            hasMargin={false}
            key="1Test description2"
          >
            Test description2
          </Paragraph>
        </div>
        <Button
          className="mfui-beta-video-block__button buttonClass"
          href="#"
        >
          Button title
        </Button>
      </div>
    </GridColumn>
    <GridColumn
      all="7"
      key="column-video"
      mobile="12"
      tablet="12"
    >
      <div
        className="mfui-beta-video-block__video-wrapper mfui-beta-video-block__video-wrapper_with-content"
      >
        <video
          autoPlay={false}
          className="mfui-beta-video-block__video"
          controls={true}
          loop={true}
          muted={true}
        >
          <source
            src="video.mp4"
            type="video/mp4"
          />
        </video>
      </div>
    </GridColumn>
  </Grid>
</div>
`;

exports[`<VideoBlock /> it renders VideoBlock with content 1`] = `
<div
  className="mfui-beta-video-block"
>
  <Grid
    className="mfui-beta-video-block__grid"
    hAlign="center"
  >
    <GridColumn
      all="5"
      key="column-content"
      mobile="12"
      orderMobile="2"
      orderTablet="2"
      tablet="12"
    >
      <div
        className="mfui-beta-video-block__content"
      >
        <Header
          as="h3"
          className="mfui-beta-video-block__header"
          color="black"
          hAlign="inherit"
        >
          Test title
        </Header>
        <div>
          <Paragraph
            className="mfui-beta-video-block__text"
            hasMargin={false}
            key="0Test description"
          >
            Test description
          </Paragraph>
          <Paragraph
            className="mfui-beta-video-block__text"
            hasMargin={false}
            key="1Test description2"
          >
            Test description2
          </Paragraph>
        </div>
        <Button
          className="mfui-beta-video-block__button"
          href="#"
        >
          Button title
        </Button>
      </div>
    </GridColumn>
    <GridColumn
      all="7"
      key="column-video"
      mobile="12"
      tablet="12"
    >
      <div
        className="mfui-beta-video-block__video-wrapper mfui-beta-video-block__video-wrapper_with-content"
      >
        <video
          autoPlay={false}
          className="mfui-beta-video-block__video"
          controls={true}
          loop={true}
          muted={true}
        >
          <source
            src="video.mp4"
            type="video/mp4"
          />
        </video>
      </div>
    </GridColumn>
  </Grid>
</div>
`;

exports[`<VideoBlock /> it renders VideoBlock with default props 1`] = `
<div
  className="mfui-beta-video-block"
>
  <Grid
    className="mfui-beta-video-block__grid"
    hAlign="center"
  >
    <GridColumn
      all="10"
      key="column-video"
      mobile="12"
      tablet="12"
    >
      <div
        className="mfui-beta-video-block__video-wrapper"
      >
        <video
          autoPlay={false}
          className="mfui-beta-video-block__video"
          controls={true}
          loop={true}
          muted={true}
        >
          <source
            src="video.mp4"
            type="video/mp4"
          />
        </video>
      </div>
    </GridColumn>
  </Grid>
</div>
`;

exports[`<VideoBlock /> it renders VideoBlock with sound turned on 1`] = `
<div
  className="mfui-beta-video-block"
>
  <Grid
    className="mfui-beta-video-block__grid"
    hAlign="center"
  >
    <GridColumn
      all="10"
      key="column-video"
      mobile="12"
      tablet="12"
    >
      <div
        className="mfui-beta-video-block__video-wrapper"
      >
        <video
          autoPlay={false}
          className="mfui-beta-video-block__video"
          controls={true}
          loop={true}
          muted={false}
        >
          <source
            src="video"
            type="video/mp4"
          />
        </video>
      </div>
    </GridColumn>
  </Grid>
</div>
`;

exports[`<VideoBlock /> it renders VideoBlock with youtube media type 1`] = `
<div
  className="mfui-beta-video-block"
>
  <Grid
    className="mfui-beta-video-block__grid"
    hAlign="center"
  >
    <GridColumn
      all="10"
      key="column-video"
      mobile="12"
      tablet="12"
    >
      <div
        className="mfui-beta-video-block__video-wrapper"
      >
        <iframe
          allow="autoplay"
          frameBorder="0"
          height="100%"
          src="https://www.youtube.com/embed/youtube?&autoplay=0&mute=1&loop=1&rel=0&controls=0&showinfo=0e&iv_load_policy=3&playlist=youtube"
          width="100%"
        />
      </div>
    </GridColumn>
  </Grid>
</div>
`;

exports[`<VideoBlock /> it renders VideoBlock with youtube media type and autoplay 1`] = `
<div
  className="mfui-beta-video-block"
>
  <Grid
    className="mfui-beta-video-block__grid"
    hAlign="center"
  >
    <GridColumn
      all="10"
      key="column-video"
      mobile="12"
      tablet="12"
    >
      <div
        className="mfui-beta-video-block__video-wrapper"
      >
        <iframe
          allow="autoplay"
          frameBorder="0"
          height="100%"
          src="https://www.youtube.com/embed/youtube?&autoplay=1&mute=1&loop=1&rel=0&controls=0&showinfo=0e&iv_load_policy=3&playlist=youtube"
          width="100%"
        />
      </div>
    </GridColumn>
  </Grid>
</div>
`;

exports[`<VideoBlock /> it renders VideoBlock with youtube media type and sound turned on 1`] = `
<div
  className="mfui-beta-video-block"
>
  <Grid
    className="mfui-beta-video-block__grid"
    hAlign="center"
  >
    <GridColumn
      all="10"
      key="column-video"
      mobile="12"
      tablet="12"
    >
      <div
        className="mfui-beta-video-block__video-wrapper"
      >
        <iframe
          allow="autoplay"
          frameBorder="0"
          height="100%"
          src="https://www.youtube.com/embed/youtube?&autoplay=0&mute=0&loop=1&rel=0&controls=0&showinfo=0e&iv_load_policy=3&playlist=youtube"
          width="100%"
        />
      </div>
    </GridColumn>
  </Grid>
</div>
`;
